client/asset/eth/eth.go

```diff
@@ -1451,13 +1475,39 @@ func (w *assetWallet) swapGas(n int, ver uint32) (oneSwap, nSwap uint64, approve
 	// If we've approved the contract to transfer, we can get a live
 	// estimate to double check.
 
+	// The amount we can estimate and ultimately the amount we can use in a
+	// single transaction is limited by the block gas limit or the tx gas
+	// limit. Estimate the number of swaps we can safely estimate here.
+	maxN := n
+	if n > 1 && nSwap > perTxGasLimit {
+		maxN = 1
+		gas := perTxGasLimit - g.Swap
+		maxN += int(gas / g.SwapAdd)
+		// nSwap will use extra gas for every fold.
+		folds := uint64(n / maxN)
+		if n%maxN == 0 {
+			folds -= 1
+		}
+		// Assume SwapAdd is less that Swap.
+		addGas := (g.Swap - g.SwapAdd) * folds
+		nSwap += addGas
+	}
+
 	// If a live estimate is greater than our estimate from configured values,
 	// use the live estimate with a warning.
-	if gasEst, err := w.estimateInitGas(w.ctx, n, ver); err != nil {
+	gasEst, err := w.estimateInitGas(w.ctx, maxN, ver)
+	if err != nil {
 		w.log.Errorf("(%d) error estimating swap gas: %v", w.assetID, err)
-		// TODO: investigate "gas required exceeds allowance".
 		return 0, 0, false, err
-	} else if gasEst > nSwap {
+	}
+	if maxN != n {
+		// If we needed to adjust the max earlier, and the estimate did
+		// not error, multiply the estimate by the fraction of swaps we
+		// were able to estimate.
+		multBy := float64(n) / float64(maxN)
+		gasEst = uint64(float64(gasEst) * multBy)
```

Comment:
I don't think this math checks, and I can't think of a way to do this without getting an additional `estimateInitGas` for the remainder. Something like...

```Go
if maxN != n {
	nFull := n / maxN
	gasEst *= uint64(nFull)
	if nRemain := n % maxN; nRemain > 0 {
		remainEst, err := w.estimateInitGas(w.ctx, nRemain, ver)
		if err != nil {
			w.log.Errorf("(%d) error estimating swap gas for remainder: %v", w.assetID, err)
			return 0, 0, false, err
		}
		gasEst += remainEst
	}
}
```

Reply:
Actually, I think if you did it this way, you can avoid the `fold` math up above.

Reply:
Oh, that works. Get another estimate for the remainder.

Reply:
Better to just recalculate for nSwap? Yeah that probably reads better.